### PR TITLE
[SPARK-22291][SQL] Conversion error when transforming array types of uuid, inet and cidr to StingType in PostgreSQL

### PR DESCRIPTION
**Pull request by jmchung**
## What changes were proposed in this pull request?

This PR fixes the conversion error when reads data from a PostgreSQL table that contains columns of `uuid[]`, `inet[]` and `cidr[]` data types. 

For example, create a table with the uuid[] data type, and insert the test data.
```SQL
CREATE TABLE users
(
    id smallint NOT NULL,
    name character varying(50),
    user_ids uuid[],
    PRIMARY KEY (id)
)

INSERT INTO users ("id", "name","user_ids") 
VALUES (1, 'foo', ARRAY
    ['7be8aaf8-650e-4dbb-8186-0a749840ecf2'
    ,'205f9bfc-018c-4452-a605-609c0cfad228']::UUID[]
)
```
Then it will throw the following exceptions when trying to load the data.
```
java.lang.ClassCastException: [Ljava.util.UUID; cannot be cast to [Ljava.lang.String;
    at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$14.apply(JdbcUtils.scala:459)
    at org.apache.spark.sql.execution.datasources.jdbc.JdbcUtils$$anonfun$14.apply(JdbcUtils.scala:458)
...
```

## How was this patch tested?

Added test in `PostgresIntegrationSuite`.

### CODE REVIEW

**Comment by jmchung** [comment]
cc @viirya Can you help review this? Thanks.

**Comment by wangyum** [comment]
Please add tests to https://github.com/apache/spark/blob/master/external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala

**Comment by viirya** [comment]
@jmchung You can add unit test into `PostgresIntegrationSuite`.

**Comment by viirya** [comment]
Seems that we don't even have test against uuid column, when you create test for uuid[], can you also create test for uuid? Thanks.

**Comment by viirya** [comment]
Besides uuid, other PostgreSQL types such as  "cidr", "inet", they are treated as StringType too, will they work?

**Comment by jmchung** [comment]
Thanks @wangyum and @viirya, I'll add the corresponding tests in `PostgresIntegrationSuite`.
To @viirya , I'm not sure if the other data types will work,  will consider them into tests, thanks!

**Comment by jmchung** [comment]
gentle ping @wangyum and @viirya, the test case has been added to `PostgresIntegrationSuite`. As @viirya mentioned early, data types `inet[]` and `cidr[]` did not work as for instance of String, they are both mapped to Java's Object class.

**Comment by viirya** [comment]
@jmchung There are another StringType types "json", "jsonb" in `PostgresDialect`. Can you also add them into tests?  Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 1]
```diff
@@ -456,8 +456,17 @@ object JdbcUtils extends Logging {
...
         case StringType =>
           (array: Object) =>
-            array.asInstanceOf[Array[java.lang.String]]
-              .map(UTF8String.fromString)
+            array match {
+              case _: Array[java.lang.String] =>
+                array.asInstanceOf[Array[java.lang.String]]
+                  .map(UTF8String.fromString)
+              case _: Array[java.util.UUID] =>
+                array.asInstanceOf[Array[java.util.UUID]]
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Can this also be covered by the following `Array[java.lang.Object]`?

**Comment by viirya** [comment]
cc @HyukjinKwon @cloud-fan Please help trigger the jenkins test if you have time. Thanks.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala [hunk 2]
```diff
@@ -18,7 +18,7 @@
 package org.apache.spark.sql.jdbc
 
 import java.sql.Connection
-import java.util.Properties
+import java.util.{Properties, UUID}
```
Comment (viirya) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: Do we use `UUID` here?

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
add to whitelist

**Comment by jmchung** [comment]
Thanks @HyukjinKwon :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 3]
```diff
@@ -456,8 +456,17 @@ object JdbcUtils extends Logging {
...
+                  .map(uuid => UTF8String.fromString(uuid.toString))
+              case _ =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: I can't currently take a close look now (it's mobile) but is it safe (or related) to allow the conversion from obj to string?

Comment (jmchung) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Will it be more appropriate to use `obj.asInstanceOf[String]` to convert?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala (hunk 2 above)
Comment (jmchung) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: Thanks! Remove the unused import statement.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 3 above)
Comment (jmchung) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: I was wrong, org.postgresql.util.PGobject cannot be cast to java.lang.String.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 1 above)
Comment (jmchung) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yes, Object can cover the UUID since it's a Superclass in Java. Should we not clearly distinguish between classes? Would you mind giving some comments? Thanks!

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Having a separate case for UUID is more clear to me and maybe easy to debug in the future. However I'm not sure how long this matching cases can be grown.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala [hunk 4]
```diff
@@ -134,11 +149,28 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationSuite {
...
+  test("SPARK-20557: column type TIMESTAMP with TIME ZONE and TIME with TIME ZONE " +
+    "should be recognized") {
     val dfRead = sqlContext.read.jdbc(jdbcUrl, "ts_with_timezone", new Properties)
     val rows = dfRead.collect()
     val types = rows(0).toSeq.map(x => x.getClass.toString)
     assert(types(1).equals("class java.sql.Timestamp"))
     assert(types(2).equals("class java.sql.Timestamp"))
   }
+
+  test("SPARK-22291: Postgresql UUID[] to Cassandra: Conversion Error") {
```
Comment (gatorsmile) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: This is not related to `Cassandra `. Please remove it from the test case name and the PR title.

Comment (jmchung) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: Thanks @gatorsmile and `Cassandra` has been removed. Feel free to correct me if the revised title is inappropriate. Do you have more comments?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 5]
```diff
@@ -456,8 +456,17 @@ object JdbcUtils extends Logging {
 
         case StringType =>
           (array: Object) =>
-            array.asInstanceOf[Array[java.lang.String]]
-              .map(UTF8String.fromString)
+            array match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: looking at how we handle StringType, we call `ResultSet.getString` even if the underling type is not string, but uuid, inet, cidr, etc. I think we shoud follow the behavior here, i.e. just call `array.asInstanceOf[Array[java.lang.Object]].map(obj => UTF8String.fromString(obj.toString))`

**Comment by cloud-fan** [comment]
LGTM except one comment, good catch!

**Comment by viirya** [comment]
This is not just for UUID[] for now. We should update the PR title.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Ok. Then we remove this UUID pattern and only have used `Array[java.lang.Object]`.

**Comment by jmchung** [comment]
Thanks @viirya, the title has been changed. Please correct me if the modified title is still inappropriate.

**Comment by jmchung** [comment]
gentle ping @cloud-fan and @viirya, there are some feedbacks about the behavior of obj to string.

``` scala
case StringType =>
  (array: Object) =>
    array.asInstanceOf[Array[java.lang.Object]]
      .map(obj => UTF8String.fromString(obj.toString))
```
As @HyukjinKwon mentioned before, it may unsafe to covnert the object to string directly, we'll have the `java.lang.NullPointerException` in test cases. `StringType` is different from other types use `nullSafeConvert` method to deal with the null, it process within `UTF8String.fromString`.

The `UTF8String.fromString` returns null and UTF8String so that we cannot use `String.valueOf(obj)` to avoid the NPE, it will raise another errors (because `null != "null"`):

```
- Type mapping for various types *** FAILED ***
  Array("a", "null", "b") did not equal List("a", null, "b") (PostgresIntegrationSuite.scala:120)
 ``` 

IMHO, if we don't want to make separation of match cases, we may remain null to meet the original design, e.g.,

``` scala
case StringType =>
  (array: Object) =>
    array.asInstanceOf[Array[java.lang.Object]]
      .map(obj => UTF8String.fromString(
        if (obj == null) null else obj.toString))
```

Alternative is we keep the String and Object cases?

Would you mind give some comments? I can keep working on this :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 6]
```diff
@@ -456,8 +456,10 @@ object JdbcUtils extends Logging {
 
         case StringType =>
           (array: Object) =>
-            array.asInstanceOf[Array[java.lang.String]]
-              .map(UTF8String.fromString)
+            // some underling types are not String such as uuid, inet, cidr, etc.
+            array.asInstanceOf[Array[java.lang.Object]]
+              .map(obj => UTF8String.fromString(
+                if (obj == null) null else obj.toString))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: nit: `if (obj == null) null else UTF8String.fromString(obj.toString)`

Comment (jmchung) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Thanks @cloud-fan !

**Comment by cloud-fan** [comment]
LGTM

**Comment by HyukjinKwon** [comment]
LGTM too

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala [hunk 7]
```diff
@@ -134,11 +149,28 @@ class PostgresIntegrationSuite extends DockerJDBCIntegrationSuite {
...
+  test("SPARK-20557: column type TIMESTAMP with TIME ZONE and TIME with TIME ZONE " +
+    "should be recognized") {
     val dfRead = sqlContext.read.jdbc(jdbcUrl, "ts_with_timezone", new Properties)
     val rows = dfRead.collect()
     val types = rows(0).toSeq.map(x => x.getClass.toString)
     assert(types(1).equals("class java.sql.Timestamp"))
     assert(types(2).equals("class java.sql.Timestamp"))
   }
+
+  test("SPARK-22291: PostgreSQL UUID[] to StringType: Conversion Error") {
```
Comment (viirya) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: Let's change this test title too.

**Comment by viirya** [comment]
LGTM

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala (hunk 7 above)
Comment (jmchung) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala]: Oops, I forgot it, thanks!!

**Comment by cloud-fan** [comment]
the last commit just change the test name and shouldn't break pyspark tests, I'm merging to master, thanks!

**Comment by cloud-fan** [comment]
@jmchung can you send a new PR for 2.2? thanks!
